### PR TITLE
Added Brave Search box on private window

### PR DESCRIPTION
**Pull request by nullhook**
Resolves https://github.com/brave/brave-browser/issues/22566

Implemented the new private ntp window which adds a search box and redirects to Brave search page when the user enters a query. We do not show suggestions dropdown while the user is typing in the search box on private windows.

Introduced two new profile prefs for the disclaimer dialog. This is to persist user's choice of hiding the expanded disclaimer dialog in the subsequent invocation of the private window and private window with Tor. Initially, the dialog is always expanded.

~**Please note:** This PR does not remove the existing private window WebUI rather adds another entry point in webui factory controller. Perhaps, we can remove the current WebUI and it's related artifacts after merge as it's a more involved task.~

~The following would need to be cleaned up:~
1. ~`components/brave_new_tab_ui/containers/privateTab`~
2. ~all private tab container imports in app.tsx~
3. ~TorObserver from `BraveNewTabMessageHandler`~
4. Private window strings should be removed from existing NewTab component grdp and moved to it's own grdp

image

image

image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

**Search test**
Input a query in the search box, press enter or click on the search icon in both private window and private window with Tor. This should redirect to Brave search.

Private window search query should go to: `https://search.brave.com/search?q={query}`
Tor window search query should go to: `https://search.brave4u7jddbv7cyviptqjc7jusxh72uik7zt6adtckl5f4nwy2v72qd.onion/search?q={query}`

**Disclaimer dialog test**
The disclaimer dialog should not be expanded by default, once you hide it by pressing "x"

**Tor observer test**
Create a new private window with Tor and see if the badge on the top left is in loading state and eventually completed with a different UI state. You should also reload the browser page after completed and/or during Tor is in progress. This will ensure if the UI state is in-sync with Tor connection.

**Guest window test**
All of the above behavior and designs should match in Guest windows as well

### CODE REVIEW

File: components/brave_new_tab_ui/brave_new_tab_searchbox.mojom
Comment (petemill) [inline components/brave_new_tab_ui/brave_new_tab_searchbox.mojom:8]: I don't think you need another interface here, the PageHandler can just be bound automatically. See how brave_news.mojom is bound.

File: components/brave_new_tab_ui/brave_new_tab_searchbox.mojom
Comment (petemill) [inline components/brave_new_tab_ui/brave_new_tab_searchbox.mojom:6]: how about we centralize this to a small component at `components/brave_search` and give it a spiffy name, like `brave_search.mojom.SearchHandler`, instead of tying to Private NTP

Comment (nullhook) [reply components/brave_new_tab_ui/brave_new_tab_searchbox.mojom]: i have added a seperate mojom for private tab, ptal

File: browser/ui/webui/brave_web_ui_controller_factory.cc
Comment (simonhong) [inline browser/ui/webui/brave_web_ui_controller_factory.cc:118]: It seems guest window still uses old private NTP.
I think it also should use new private NTP?

**Review by simonhong — COMMENTED** [review]
It looks like new private NTP doesn't handle DDG toggle button?
https://github.com/brave/brave-browser/issues/23091#issuecomment-1147509208

File: browser/ui/webui/brave_web_ui_controller_factory.cc
Comment (nullhook) [reply browser/ui/webui/brave_web_ui_controller_factory.cc]: added guest window flag

**Comment by nullhook** [comment]
> It looks like new private NTP doesn't handle DDG toggle button? [brave/brave-browser#23091 (comment)](https://github.com/brave/brave-browser/issues/23091#issuecomment-1147509208)

Yes, the new private NTP doesn't handle DDG toggle button

File: components/brave_private_new_tab_ui/resources/page/private_new_tab.html
Comment (fmarier) [inline components/brave_private_new_tab_ui/resources/page/private_new_tab.html:12]: I don't know whether we've thought about this yet, but in general it would be good to avoid inline scripts or inline styles so that we can use a strong CSP policy for these pages in the future.

So my recommendation here would be to move this to a `/styles.css` like the `/strings.js` we already have.

File: components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx
Comment (fmarier) [inline components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx:25]: At the moment, the most important part of the Tor message ("[...]] you should be using Tor browser instead.") is below the fold.

Can we make it so that it shows up in full when the screen is large enough?

**Review by fmarier — CHANGES_REQUESTED** [review]

File: browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.h
Comment (simonhong) [inline browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.h:73]: `raw_ptr`

**Review by simonhong — APPROVED** [review]
++ 👍🏼 
@rebron We decided to delete DDG toggle in private NTP?

File: components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx
Comment (petemill) [reply components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx:25]: @nullhook perhaps make the max-height something like `(100vh - margin - approximate search box height) / 2`?

File: components/brave_private_new_tab_ui/resources/page/private_new_tab.html
Comment (petemill) [inline components/brave_private_new_tab_ui/resources/page/private_new_tab.html:13]: Seems like we're selecting `#root` here but using `#mountPoint` in the html.

File: components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx
Comment (petemill) [inline components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx:127]: Also needs accessible label

**Review by petemill — CHANGES_REQUESTED** [review]
Great work, looks really nice. Just some minor issues to look at, I think.

Also perhaps we should look at focus order. I expect when I open a new tab that the focus goes to the address bar. That works still. But then when I move focus to the web content (a single Tab press for me), I expect the search box to receive focus. However, the (i) button seems to have first focus. Perhaps we need to set tab index or just autofocus on the search box input.

File: browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.cc
Comment (nullhook) [reply browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.cc:101]: If page_ isn't bound then DCHECK would be called. I'm not sure if another check is required here? Perhaps, time and effort in writing tests for this can ensure that it won't escape into production.

File: components/brave_private_new_tab_ui/resources/page/components/search-box/index.tsx
Comment (nullhook) [reply components/brave_private_new_tab_ui/resources/page/components/search-box/index.tsx:71]: yes, this is ran through the optimizer

**Comment by nullhook** [comment]
> Great work, looks really nice. Just some minor issues to look at, I think.
> 
> Also perhaps we should look at focus order. I expect when I open a new tab that the focus goes to the address bar. That works still. But then when I move focus to the web content (a single Tab press for me), I expect the search box to receive focus. However, the (i) button seems to have first focus. Perhaps we need to set tab index or just autofocus on the search box input.

A tab index isn't needed to solve or even autofocus. The ordering of html needed to be changed. My current order is:

```
<Badge />
<DialogBox />
<SearchBox />
```

I need to swap the last two. As the dialog box is out of normal document flow, It's an absolutely positioned.

File: components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx
Comment (nullhook) [reply components/brave_private_new_tab_ui/resources/page/components/disclaimer-dialog/index.tsx]: added this in css so the browser can do re-computes. Updated screenshot as well @fmarier - On screen size with space you won't see scroll anymore.

File: components/brave_private_new_tab_ui/resources/page/hooks.ts
Comment (nullhook) [inline components/brave_private_new_tab_ui/resources/page/hooks.ts:29]: @petemill When you close the disclaimer box for the first time, reload the page, there's a flash of the disclaimer box on subsequent reloads. Unless, if you destroy the window and open it again.

I wonder if the default here should always be false which matches with the default of user pref?

File: components/brave_private_new_tab_ui/common/pref_names.h
Comment (petemill) [inline components/brave_private_new_tab_ui/common/pref_names.h:13]: I think the naming of these constants has been confusing me - how about we match to the intention and the string? Something like `kBravePrivateWindowDisclaimerDismissed`?

File: components/brave_private_new_tab_ui/resources/page/hooks.ts
Comment (petemill) [reply components/brave_private_new_tab_ui/resources/page/hooks.ts:29]: Well, should it be the other way round to avoid this? In other words, do not show the "disclaimer" by default. And then if the API returns the decision that we should show it then we show it. Better to have it appear slightly after page load than flash visible then hidden, right? And then you can probably just remove the loadTimeData property. 

For times where we really do want this to be available at load-time and not wait for an async mojom call, there is a way around this: using `render_frame_host->SetWebUIProperty` in the Browser and `chrome.getVariableValue` in JS. And I believe that's updateable on a refresh. But I don't suggest it for this case.

**Review by petemill — CHANGES_REQUESTED** [review]
The focus is mostly fixed, except from when the disclaimer element is showing and I refresh the page. Then the focus is on the close button, when I would expect it to be the input for the search query. What do you think?

File: components/brave_private_new_tab_ui/common/pref_names.h
Comment (nullhook) [reply components/brave_private_new_tab_ui/common/pref_names.h]: makes sense

**Comment by nullhook** [comment]
The focus issue should be fixed now.

File: components/brave_private_new_tab_ui/resources/page/hooks.ts
Comment (nullhook) [inline components/brave_private_new_tab_ui/resources/page/hooks.ts:39]: @petemill I hope this is clear

Comment (petemill) [reply components/brave_private_new_tab_ui/resources/page/hooks.ts:39]: If all we're trying to do is not display components until we've fetched data for the component, then I suggest we just do that. In other words, initialize state properties to `null` (or `undefined`) for API calls which fetch data, and when the data is in that initial state then do not show the relevant UI.

File: components/brave_private_new_tab_ui/resources/page/hooks.ts
Comment (petemill) [inline components/brave_private_new_tab_ui/resources/page/hooks.ts:18]: nit: you don't need to re-define type for `res` here (and other calls to similar functions in this module) since the type is coming from mojom's JSDoc. Might be a bit clearer to read without them.

File: resources/resource_ids.spec
Comment (petemill) [inline resources/resource_ids.spec:187]: nit: I don't think this webui will be using more than 1 ID (no external images or dynamic JS modules), so reserving 250 seems extreme. Perhaps put this to 20 at most.

**Review by petemill — CHANGES_REQUESTED** [review]
From a UI perspective this is great now! Please get the CI all green.
One more nit I have though is that I think user expectation on any search or address bar is that hitting cmd-enter after typing text would open the results page in a new tab - can you add that?
It would also be great to see the existing private NTP code removed from the binary in this PR as this PR also makes it not used at all.
And then I think we need the browser test as part of the security review to verify that the tab opened via the tor NTP still opens the results page in a tor context?

**Review by fmarier — APPROVED** [review]
The high-level changes look good to me.

File: resources/resource_ids.spec
Comment (nullhook) [reply resources/resource_ids.spec]: makes sense

**Comment by nullhook** [comment]
I have added browsertest and `cmd/ctrl+enter` event

File: browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_browsertest.cc
Comment (petemill) [inline browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_browsertest.cc:52]: Maybe refer to selector and not Id

Comment (nullhook) [reply browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_browsertest.cc]: done

**Comment by petemill** [comment]
> Private window strings should be removed from existing NewTab component grdp and moved to it's own grdp

We can wait for that before merge?

File: browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.cc
Comment (petemill) [inline browser/ui/webui/private_new_tab_page/brave_private_new_tab_page_handler.cc:104]: 👍

**Review by petemill — APPROVED** [review]
Great work, please make sure to file anything you decided to do as follow-ups

**Comment by rebron** [comment]
Great job @nullhook! Looks great.

Filed https://github.com/brave/brave-browser/issues/23732 to update the logo to replace with normal logo (without the beta label) as we're out of beta with Brave Search.

**Pull request merged**
